### PR TITLE
Derive word-confidence boundaries from the decoded text itself

### LINKED ISSUE #15143: Mismatch of len(words) and len(word_confidence) in parakeet-tdt-0.6b-v3 transcription
**Describe the bug**

I am running a transcription pipeline using the parakeet-tdt-0.6b-v3 model. In order to not run out of memory, I have implemented an algorithm for chunking longer files into shorter chunks, run the transcription on the chunks, then reassemble the transcription of the chunks to obtain the transcription of the file. 
The chunking is done by running the "pyannote/voice-activity-detection" VAD model on the audio, then aggregating the segments with active voice until a given maximum length (e.g. 10 min) is exceeded. The audio is then cut in the mid point between the end of one active segment and the start of the next active segment to prevent the audio being cut in the middle of a word. 
I have tried using a naive chunking (without checking for voice activity) before, which lead to the same error occuring. 

I have tested this on publicly available audio recordings of the german parliament. 
This works well for most files I tested. But for some, I run into the following RuntimeError in line 463 of `.venv/lib/python3.12/site-packages/nemo/collections/asr/parts/utils/asr_confidence_utils.py `: 
```
---------------------------------------------------------------------------
RuntimeError                              Traceback (most recent call last)
Cell In[7], [line 60](vscode-notebook-cell:?execution_count=7&line=60)
     58 # 4) ASR
     59 with torch.inference_mode(), torch.amp.autocast("cuda", dtype=torch.float16):
---> [60](vscode-notebook-cell:?execution_count=7&line=60)     out = asr_model.transcribe(
     61         chunk_arrays, 
     62         batch_size=batch_size,
     63         num_workers=4, 
     64         return_hypotheses=True
     65     )

File ~/repos/nvidia-parakeet-testing/.venv/lib/python3.12/site-packages/torch/utils/_contextlib.py:120, in context_decorator.<locals>.decorate_context(*args, **kwargs)
    117 @functools.wraps(func)
    118 def decorate_context(*args, **kwargs):
    119     with ctx_factory():
--> [120](https://vscode-remote+ssh-002dremote-002bpdeki01.vscode-resource.vscode-cdn.net/home/sfelix/repos/nvidia-parakeet-testing/~/repos/nvidia-parakeet-testing/.venv/lib/python3.12/site-packages/torch/utils/_contextlib.py:120)         return func(*args, **kwargs)

File ~/repos/nvidia-parakeet-testing/.venv/lib/python3.12/site-packages/nemo/collections/asr/models/rnnt_models.py:308, in EncDecRNNTModel.transcribe(self, audio, batch_size, return_hypotheses, partial_hypothesis, num_workers, channel_selector, augmentor, verbose, timestamps, override_config)
    304             self.cfg.decoding.preserve_alignments = False
    306     self.change_decoding_strategy(self.cfg.decoding, verbose=False)
--> [308](https://vscode-remote+ssh-002dremote-002bpdeki01.vscode-resource.vscode-cdn.net/home/sfelix/repos/nvidia-parakeet-testing/~/repos/nvidia-parakeet-testing/.venv/lib/python3.12/site-packages/nemo/collections/asr/models/rnnt_models.py:308) return super().transcribe(
    309     audio=audio,
    310     batch_size=batch_size,
    311     return_hypotheses=return_hypotheses,
    312     num_workers=num_workers,
    313     channel_selector=channel_selector,
    314     augmentor=augmentor,
    315     verbose=verbose,
    316     timestamps=timestamps,
    317     override_config=override_config,
    318     # Additional arguments
    319     partial_hypothesis=partial_hypothesis,
    320 )

File ~/repos/nvidia-parakeet-testing/.venv/lib/python3.12/site-packages/torch/utils/_contextlib.py:120, in context_decorator.<locals>.decorate_context(*args, **kwargs)
    117 @functools.wraps(func)
    118 def decorate_context(*args, **kwargs):
    119     with ctx_factory():
--> [120](https://vscode-remote+ssh-002dremote-002bpdeki01.vscode-resource.vscode-cdn.net/home/sfelix/repos/nvidia-parakeet-testing/~/repos/nvidia-parakeet-testing/.venv/lib/python3.12/site-packages/torch/utils/_contextlib.py:120)         return func(*args, **kwargs)

File ~/repos/nvidia-parakeet-testing/.venv/lib/python3.12/site-packages/nemo/collections/asr/parts/mixins/transcription.py:270, in TranscriptionMixin.transcribe(self, audio, batch_size, return_hypotheses, num_workers, channel_selector, augmentor, verbose, timestamps, override_config, **config_kwargs)
    267 try:
    268     generator = self.transcribe_generator(audio, override_config=transcribe_cfg)
--> [270](https://vscode-remote+ssh-002dremote-002bpdeki01.vscode-resource.vscode-cdn.net/home/sfelix/repos/nvidia-parakeet-testing/~/repos/nvidia-parakeet-testing/.venv/lib/python3.12/site-packages/nemo/collections/asr/parts/mixins/transcription.py:270)     for processed_outputs in generator:
    271         # Store results
    272         if isinstance(processed_outputs, list):
    273             # Create a results of the same type as each element in processed_outputs
    274             if results is None:

File ~/repos/nvidia-parakeet-testing/.venv/lib/python3.12/site-packages/nemo/collections/asr/parts/mixins/transcription.py:370, in TranscriptionMixin.transcribe_generator(self, audio, override_config)
    368 # Run forward pass
    369 model_outputs = self._transcribe_forward(test_batch, transcribe_cfg)
--> [370](https://vscode-remote+ssh-002dremote-002bpdeki01.vscode-resource.vscode-cdn.net/home/sfelix/repos/nvidia-parakeet-testing/~/repos/nvidia-parakeet-testing/.venv/lib/python3.12/site-packages/nemo/collections/asr/parts/mixins/transcription.py:370) processed_outputs = self._transcribe_output_processing(model_outputs, transcribe_cfg)
    372 # clear up memory
    373 del test_batch, model_outputs

File ~/repos/nvidia-parakeet-testing/.venv/lib/python3.12/site-packages/nemo/collections/asr/models/rnnt_models.py:946, in EncDecRNNTModel._transcribe_output_processing(self, outputs, trcfg)
    943 encoded = outputs.pop('encoded')
    944 encoded_len = outputs.pop('encoded_len')
--> [946](https://vscode-remote+ssh-002dremote-002bpdeki01.vscode-resource.vscode-cdn.net/home/sfelix/repos/nvidia-parakeet-testing/~/repos/nvidia-parakeet-testing/.venv/lib/python3.12/site-packages/nemo/collections/asr/models/rnnt_models.py:946) hyp = self.decoding.rnnt_decoder_predictions_tensor(
    947     encoded,
    948     encoded_len,
    949     return_hypotheses=trcfg.return_hypotheses,
    950     partial_hypotheses=trcfg.partial_hypothesis,
    951 )
    952 # cleanup memory
    953 del encoded, encoded_len

File ~/repos/nvidia-parakeet-testing/.venv/lib/python3.12/site-packages/nemo/collections/asr/parts/submodules/rnnt_decoding.py:640, in AbstractRNNTDecoding.rnnt_decoder_predictions_tensor(self, encoder_output, encoded_lengths, return_hypotheses, partial_hypotheses)
    635 if return_hypotheses:
    636     # greedy decoding, can get high-level confidence scores
    637     if self.preserve_frame_confidence and (
    638         self.preserve_word_confidence or self.preserve_token_confidence
    639     ):
--> [640](https://vscode-remote+ssh-002dremote-002bpdeki01.vscode-resource.vscode-cdn.net/home/sfelix/repos/nvidia-parakeet-testing/~/repos/nvidia-parakeet-testing/.venv/lib/python3.12/site-packages/nemo/collections/asr/parts/submodules/rnnt_decoding.py:640)         hypotheses = self.compute_confidence(hypotheses)
    641     return hypotheses
    643 return [Hypothesis(h.score, h.y_sequence, h.text) for h in hypotheses]

File ~/repos/nvidia-parakeet-testing/.venv/lib/python3.12/site-packages/nemo/collections/asr/parts/submodules/rnnt_decoding.py:770, in AbstractRNNTDecoding.compute_confidence(self, hypotheses_list)
    768 if self.preserve_word_confidence:
    769     for hyp in hypotheses_list:
--> [770](https://vscode-remote+ssh-002dremote-002bpdeki01.vscode-resource.vscode-cdn.net/home/sfelix/repos/nvidia-parakeet-testing/~/repos/nvidia-parakeet-testing/.venv/lib/python3.12/site-packages/nemo/collections/asr/parts/submodules/rnnt_decoding.py:770)         hyp.word_confidence = self._aggregate_token_confidence(hyp)
    771 return hypotheses_list

File ~/repos/nvidia-parakeet-testing/.venv/lib/python3.12/site-packages/nemo/collections/asr/parts/submodules/rnnt_decoding.py:1799, in RNNTBPEDecoding._aggregate_token_confidence(self, hypothesis)
   1787 def _aggregate_token_confidence(self, hypothesis: Hypothesis) -> List[float]:
   1788     """
   1789     Implemented by subclass in order to reduce token confidence to a word-level confidence.
   1790 
   (...)   1797         A list of word-level confidence scores.
   1798     """
-> [1799](https://vscode-remote+ssh-002dremote-002bpdeki01.vscode-resource.vscode-cdn.net/home/sfelix/repos/nvidia-parakeet-testing/~/repos/nvidia-parakeet-testing/.venv/lib/python3.12/site-packages/nemo/collections/asr/parts/submodules/rnnt_decoding.py:1799)     return self._aggregate_token_confidence_subwords_sentencepiece(
   1800         hypothesis.words, hypothesis.token_confidence, hypothesis.y_sequence
   1801     )

File ~/repos/nvidia-parakeet-testing/.venv/lib/python3.12/site-packages/nemo/collections/asr/parts/utils/asr_confidence_utils.py:464, in ConfidenceMixin._aggregate_token_confidence_subwords_sentencepiece(self, words, token_confidence, token_ids)
    462         word_confidence.append(self._aggregate_confidence(token_confidence[j : len(token_ids)]))
    463 if len(words) != len(word_confidence):
--> [464](https://vscode-remote+ssh-002dremote-002bpdeki01.vscode-resource.vscode-cdn.net/home/sfelix/repos/nvidia-parakeet-testing/~/repos/nvidia-parakeet-testing/.venv/lib/python3.12/site-packages/nemo/collections/asr/parts/utils/asr_confidence_utils.py:464)     raise RuntimeError(
    465         f"""Something went wrong with word-level confidence aggregation.\n
    466     Please check these values for debugging:\n
    467     len(words): {len(words)},\n
    468     len(word_confidence): {len(word_confidence)},\n
    469     recognized text: `{' '.join(words)}`"""
    470     )
    471 return word_confidence

RuntimeError: Something went wrong with word-level confidence aggregation.

            Please check these values for debugging:

            len(words): 1201,

            len(word_confidence): 1202,

            recognized text: `Beratung des Antrags der Fraktion DIE LINKE mit dem Titel Mindestbetrag des Elterngeldes erhöhen Familien und ihre [...] `
```
The recognized text in the error message matches the text spoken in the audio chunk.
In every instance of this Error occuring, there were more word_confidences than words. 
I have looked into the chunks created by my preprocessing and everything seems to be working as intended. There are no chunks starting or ending in the middle of a word.

**Steps/Code to reproduce bug**

Here is the code I ran to obtain the error. The helper functions at the start are just for chunking and loading the audio with torchaudio.

The audio file `plenum_7_31min.wav` is publicly available here: https://www.bundestag.de/mediathek/video?videoid=7637305
(Direct download link: http://cldf-od.r53.cdn.tv1.eu/1000153copo/ondemand/app144277506/145293313/7637305/7637305_mp3_128kb_stereo_de_128.mp3?fdl=1)

```python
import torch
import torchaudio
import os
import nemo.collections.asr as nemo_asr
from nemo.collections.asr.parts.utils.asr_confidence_utils import (
    ConfidenceConfig, ConfidenceMethodConfig
)
from nemo.collections.asr.parts.submodules.ctc_decoding import CTCDecodingConfig
from pyannote.audio import Pipeline

torch.backends.cuda.matmul.allow_tf32 = True
torch.backends.cudnn.allow_tf32 = True
torch.backends.cudnn.benchmark = True

### HELPER / UTILITY FUNCTIONS

def load_audio(path, target_sr=16000):
    wav, sr = torchaudio.load(path)          # [C, T]
    wav = wav.mean(0, keepdim=False)         # mono [T]
    if sr != target_sr:
        wav = torchaudio.functional.resample(wav, sr, target_sr)
        sr = target_sr
    return wav.numpy(), sr


def vad_build_chunks_gap_mid_capped(vad_annotation,
                                    max_chunk_s: float,
                                    audio_duration_s: float,
                                    gap_safety: float = 0.05):
    """
    Build chunks whose bounds are gap midpoints, but clamp the end so that
    (end - start) <= max_chunk_s while staying inside the current gap.

    - start := midpoint of gap before first included VAD segment
    - end   := midpoint of gap after last included VAD segment (or clamped earlier)
    - gap_safety keeps cuts strictly inside gaps by ±gap_safety seconds
    """
    # collect speech segments
    segs = [(float(s.start), float(s.end)) for s in vad_annotation.get_timeline().support()]
    segs.sort()
    if not segs:
        return []

    def prev_gap_mid(idx: int) -> float:
        if idx == 0:
            # gap is [0, segs[0].start]
            return 0.5 * (0.0 + segs[0][0])
        # gap is [segs[idx-1].end, segs[idx].start]
        return 0.5 * (segs[idx-1][1] + segs[idx][0])

    def next_gap_mid(idx: int) -> float:
        if idx == len(segs) - 1:
            # gap is [segs[-1].end, audio_end]
            return 0.5 * (segs[-1][1] + audio_duration_s)
        # gap is [segs[idx].end, segs[idx+1].start]
        return 0.5 * (segs[idx][1] + segs[idx+1][0])

    chunks = []
    i, n = 0, len(segs)

    while i < n:
        # start at midpoint of gap before seg i
        start = prev_gap_mid(i)

        last_end = segs[i][1]
        j = i

        while True:
            # candidate end if we cut after seg j at the next gap midpoint
            nat_end = next_gap_mid(j)
            span = nat_end - start

            if span <= max_chunk_s:
                # try to include next segment if any
                if j + 1 < n:
                    j += 1
                    last_end = segs[j][1]
                    continue
                else:
                    # no more segments; we will end after seg j
                    pass

            # We are finalizing chunk after seg j.
            # First, possibly back off if the following gap is too narrow
            # to place a cut inside it while respecting max_chunk_s.
            while True:
                if j < n - 1:
                    gap_lo, gap_hi = last_end, segs[j + 1][0]      # gap where we cut
                else:
                    gap_lo, gap_hi = last_end, audio_duration_s    # last gap to audio end

                earliest_end = gap_lo + gap_safety
                max_allowed_end = start + max_chunk_s

                # If we cannot place the cut inside the gap without exceeding
                # max_chunk_s, and we still have more than one segment in this
                # chunk, drop the last one from this chunk and retry.
                if earliest_end > max_allowed_end and j > i:
                    j -= 1
                    last_end = segs[j][1]
                    continue
                break

            # recompute nat_end for (possibly updated) j
            nat_end = next_gap_mid(j)
            if j < n - 1:
                gap_lo, gap_hi = last_end, segs[j + 1][0]
            else:
                gap_lo, gap_hi = last_end, audio_duration_s

            # desired cap-respecting end (no later than start + max_chunk_s)
            cap_end = min(nat_end, start + max_chunk_s)

            # keep inside gap with safety margins, without breaking the max cap
            # at this point we know gap_lo + gap_safety <= start + max_chunk_s
            end = max(cap_end, gap_lo + gap_safety)
            end = min(end, gap_hi - gap_safety)
            end = max(end, start + 1e-4)  # numerical guard

            # max-length guarantee
            if end - start > max_chunk_s + 1e-6:
                # should not happen under the assumption that no single segment
                # is longer than max_chunk_s; clamp as a last resort
                end = start + max_chunk_s

            chunks.append((start, end))

            # next chunk starts at seg j+1
            i = j + 1
            break

    # coalesce numerically contiguous chunks without breaking max_chunk_s
    merged = []
    tol = 1e-3
    eps = 1e-6
    for s, e in chunks:
        if not merged:
            merged.append([s, e])
        else:
            ps, pe = merged[-1]
            if (s - pe < tol) and ((e - ps) <= max_chunk_s + eps):
                # merge without exceeding max_chunk_s
                merged[-1][1] = max(pe, e)
            else:
                merged.append([s, e])

    return [(s, e) for s, e in merged if e > s + eps]

### RUN PIPELINE

os.environ["CUDA_VISIBLE_DEVICES"] = "1"

asr_model = nemo_asr.models.EncDecCTCModel.from_pretrained(model_name="nvidia/parakeet-tdt-0.6b-v3")
# Set attention mechanism to local sliding window; more efficient and accurate for long audios
asr_model.change_attention_model(self_attention_model="rel_pos_local_attn", att_context_size=[256, 256])
method = ConfidenceMethodConfig(name="entropy", entropy_type="gibbs", entropy_norm="lin")
# enable timestamps and word confidences
conf = ConfidenceConfig(
    preserve_frame_confidence=False,
    preserve_token_confidence=False,
    preserve_word_confidence=True,
    exclude_blank=True, 
    aggregation="mean",
    method_cfg=method
)
asr_model.change_decoding_strategy(
    CTCDecodingConfig(
        compute_timestamps=True,
        ctc_timestamp_type="word",
        preserve_alignments=True,
        confidence_cfg=conf
    )
)
asr_model = asr_model.eval().to("cuda")
asr_model.preprocessor.to("cuda")

HF_TOKEN = "#####"
vad_pipeline = Pipeline.from_pretrained(
    "pyannote/voice-activity-detection",
    use_auth_token=HF_TOKEN
)
vad_pipeline.to(torch.device("cuda"))

test_audio = "./test_files/bt/out/plenum_7_31min.wav"
max_chunk_s = 600
batch_size = 8

wav, sr = load_audio(test_audio, 16000)
duration_s = len(wav) / sr

# 1) VAD
torch.cuda.reset_peak_memory_stats()
vad = vad_pipeline(test_audio)
timeline = list(vad.get_timeline().support())

# 2) VAD-aligned chunks (simple)
chunks = vad_build_chunks_gap_mid_capped(vad, max_chunk_s=max_chunk_s, audio_duration_s=duration_s)

# 3) Prepare inputs + offsets
chunk_arrays, offsets_sec = [], []
for (s, e) in chunks:
    lo = int(round(s * sr)); hi = int(round(e * sr))
    arr = wav[lo:hi].copy()
    chunk_arrays.append(arr)
    offsets_sec.append(s)

# 4) ASR
with torch.inference_mode(), torch.amp.autocast("cuda", dtype=torch.float16):
    out = asr_model.transcribe(
        chunk_arrays, 
        batch_size=batch_size,
        num_workers=4, 
        return_hypotheses=True
    )
```

**Expected behavior**

Output as Hypotheses with the transcription, timestamps and confidences of the chunks for postprocessing.
This works for multiple other files I tested. 

**Environment overview (please complete the following information)**

 - Environment location: via SSH on my company's workstation, running locally on GPU
 
 - Method of NeMo install: uv .venv with the following requirements.txt:
 ```
nemo_toolkit[all]==2.4.0
cuda-python>=12.3
nvidia-ml-py
jupyter
ipywidgets
ipykernel
soundfile
faster-whisper
torchaudio<2.9
torchcodec
pyannote-audio
```

**Environment details**

If NVIDIA docker image is used you don't need to specify these.
Otherwise, please provide:
- OS version: Ubuntu 24.04.3 LTS, 
- PyTorch version: 2.8.0+cu128
- Python version: 3.12.11

**Additional context**

GPU is the Quadro RTX 6000. 
I have manually checked the chunk breakpoints of the audio produced by the VAD, but cannot find anything suspicious in the audio or the speech therein.

### PR DESCRIPTION
**Pull request by Eoin-Houstoun**
# What does this PR do ?

Fixes the `RuntimeError: Something went wrong with word-level confidence aggregation` (`len(words) != len(word_confidence)`) crash in subword confidence aggregation with parakeet-tdt-0.6b-v3. Fixes #15143.

**Collection**: ASR

# Changelog

- `ConfidenceMixin._aggregate_token_confidence_subwords_sentencepiece`: group tokens into words using the decode-and-split operation itself as the boundary oracle, so the group count matches `words = text.split()` by construction; an O(n) fast path handles the common case with a single `decode_ids_to_tokens` call and the oracle only runs when the fast path disagrees.
- Add tests covering word-initial punctuation pieces, byte-fallback whitespace, `<unk>` adjacency, and count invariants at scale.

# Details

`words` comes from splitting the fully decoded text, but word confidence was aggregated by re-deriving boundaries with a per-piece heuristic (a piece starting with `▁` opens a word). The v3 multilingual vocabulary breaks that assumption two ways: punctuation-initial pieces (`▁-`, `▁¿`, standalone `▁`) where the decoded text has its space stripped, and byte-fallback whitespace (U+00A0, U+2009, common in German like `z. B.`) that splits the text without any `▁` marker. Each occurrence puts the heuristic one word out of step with `text.split()`, which is exactly the off-by-small mismatches in the report (1202 vs 1201). The prior fixes in #15128 and #15249 addressed a benchmarking helper and two other mismatch sources respectively; the boundary heuristic itself was untouched, which is why the reporter still hits this on 2.7.0.

Consistency of `len(words) == len(word_confidence)` on the v3 tokenizer, before and after:

| Text case | main | this PR |
|---|---|---|
| unpunctuated controls (2) | consistent | consistent |
| "Für die CDU-CSU-Fraktion..." (reporter's fragment, unspaced) | consistent | consistent |
| comma text | consistent | consistent |
| "Für die CDU - CSU" (spaced dash) | RuntimeError | consistent |
| spaced dashes mid-sentence | RuntimeError | consistent |
| "Hola¿Qué tal?" (inverted marks) | RuntimeError | consistent |

The fix was found by Artemis Discovery (TurinTech) working from this issue's text, scored on exactly that consistency check (57 on main, 100 with this change). I reviewed every line, and the byte-fallback-whitespace failure family in the code comment was its finding, not mine.

One known approximation worth stating: when the oracle path reconciles a count difference caused by punctuation-space stripping, the confidence bucket boundaries can shift by one token around the stripped punctuation (counts are exact; the per-word attribution near such punctuation is approximate). Happy to tighten that in a follow-up if wanted.

### CODE REVIEW

File: nemo/collections/asr/parts/utils/asr_confidence_utils.py [hunk 1]
```diff
@@ -448,24 +448,93 @@ def _aggregate_token_confidence_subwords_sentencepiece(
...
+                # Slow, exact path: use the decode-and-split operation itself as the boundary oracle.
+                # This handles byte-fallback whitespace characters (non-breaking space U+00A0, thin
+                # space U+2009 — common in German such as ``z. B.`` or grouped numbers) that carry no
+                # ``▁`` marker and are therefore invisible to the fast path, yet split ``text`` into
+                # extra words. ``decode_ids_to_str`` routes through the same tokenizer path that built
+                # ``text``, so this reproduces ``text.split()`` exactly, by construction.
+                groups = []
+                prev_count = 0
+                for i in range(len(int_token_ids)):
+                    cur_count = len(self.decode_ids_to_str(int_token_ids[: i + 1]).split())
```
Comment (nithinraok) [inline nemo/collections/asr/parts/utils/asr_confidence_utils.py:510]: I think this fallback oracle may not always match the text that produced `words`. For RNNT, `hyp.text` is produced with `decode_tokens_to_str_with_strip_punctuation()`, but this line uses plain `decode_ids_to_str(prefix).split()`.

  Example:

  ```text
  pieces:        ▁CDU  ▁-  ▁CSU
  raw decode:    "CDU - CSU"
  postprocessed: "CDU- CSU"
  words:         ["CDU-", "CSU"]
```
 
The fallback sees three raw words, then clamps/reconciles to two groups. That avoids the RuntimeError, but the buckets can shift from the desired `[CDU, -] [CSU]` to `[CDU] [-, CSU]`, so the dash confidence is assigned to the wrong word.

  Could we either apply the same normalization here as the caller used to produce words, or add an exact-bucket test with distinct confidences? The current tests mostly assert only `len(wc) == len(words)`, which would not catch this attribution shift.

File: nemo/collections/asr/parts/utils/asr_confidence_utils.py [hunk 2]
```diff
@@ -448,24 +448,93 @@ def _aggregate_token_confidence_subwords_sentencepiece(
...
+            # ``len(words) != len(word_confidence)`` RuntimeError on some transcripts.
+            #
+            # Instead we walk the tokens left to right and track how many whitespace-delimited words
+            # the decoded prefix contains. ``decode_ids_to_str`` routes through the same tokenizer
+            # path that built ``text`` (``tokenizer.ids_to_text``), so ``prev_count`` below is exactly
+            # the running word count of ``text``. A token opens a new word whenever the decoded
+            # prefix gains a word; otherwise it continues the current word. Byte-fallback bytes that
+            # only complete into a character (or whitespace) once the next byte arrives are handled
+            # naturally, because the word count only advances once the decoded text actually gains a
+            # whitespace-delimited token.
```
Comment (nithinraok) [inline nemo/collections/asr/parts/utils/asr_confidence_utils.py:472]: reduce the comment. to just brief

**Review by nithinraok — CHANGES_REQUESTED** [review]

File: nemo/collections/asr/parts/utils/asr_confidence_utils.py (hunk 1 above)
Comment (Eoin-Houstoun) [reply nemo/collections/asr/parts/utils/asr_confidence_utils.py]: Thanks for the review - good catch, this reproduced exactly as you described.

`_aggregate_token_confidence_subwords_sentencepiece` now takes the ids-to-text call as a parameter, and `RNNTBPEDecoding` passes its own `decode_tokens_to_str_with_strip_punctuation` so the oracle matches the decode that produced `words`. Default is unchanged (`decode_ids_to_str`) for callers whose text is the plain decode.

On your `CDU / - / CSU` example, with token confidences 0.9 / 0.1 / 0.8:

```
before:  [0.9, 0.45]   # dash pooled into CSU
after:   [0.5, 0.8]    # dash stays with CDU-
```

Also added the exact-bucket test you asked for - it asserts those values with distinct confidences rather than just the length, plus one pinning the default oracle to the raw decode.

File: nemo/collections/asr/parts/utils/asr_confidence_utils.py (hunk 2 above)
Comment (Eoin-Houstoun) [reply nemo/collections/asr/parts/utils/asr_confidence_utils.py]: Trimmed - the block comments are down to two or three lines each now.

**Review by nithinraok — DISMISSED** [review]
LGTM

File: nemo/collections/asr/parts/submodules/rnnt_decoding.py [hunk 3]
```diff
@@ -1752,8 +1752,13 @@ def _aggregate_token_confidence(self, hypothesis: Hypothesis) -> List[float]:
         Returns:
             A list of word-level confidence scores.
         """
+        # ``words`` comes from ``hypothesis.text``, which strips the space before punctuation, so
+        # the boundary oracle has to apply the same transform or a token lands on the wrong word.
         return self._aggregate_token_confidence_subwords_sentencepiece(
```
Comment (nithinraok) [inline nemo/collections/asr/parts/submodules/rnnt_decoding.py:1757]: Should;t we also have similar change in ``ctc_deoding.py::_aggregate_token_confidence``

```python 
def _aggregate_token_confidence(self, hypothesis: Hypothesis) -> List[float]:
    token_ids = hypothesis.text[0]
    decode = self.decode_tokens_to_str_with_strip_punctuation

    return self._aggregate_token_confidence_subwords_sentencepiece(
        decode(token_ids).split(),
        hypothesis.token_confidence,
        token_ids,
        decode,
    )
```

Comment (Eoin-Houstoun) [reply nemo/collections/asr/parts/submodules/rnnt_decoding.py:1757]: Thanks Nithin, agreed. I have included that in this PR.

**Comment by nithinraok** [comment]
@artbataev for final approval
